File: ts/packages/sdk/src/core/cron.ts
import { CronExpressionParser } from "cron-parser";

const SHORTCUTS: Record<string, string> = {
  "@hourly": "0 * * * *",
  "@daily": "0 0 * * *",
  "@midnight": "0 0 * * *",
  "@weekly": "0 0 * * 0",
  "@monthly": "0 0 1 * *",
  "@yearly": "0 0 1 1 *",
  "@annually": "0 0 1 1 *",
};

const EVERY_DURATION = /^@every\s+(?<n>\d+)(?<unit>s|m|h)$/i;

/**
 * Validate (but do not normalize) a schedule expression.
 *
 * Returns null on success; otherwise an error message.
 *
 * The five-field cron expression and the documented shortcuts (D-004) are
 * accepted. `@every <duration>` is accepted only with whole-second
 * resolution where the resulting interval is at least one minute (the v1
 * resolution floor — see Limitation 3).
 */
export function validateSchedule(expr: string): string | null {
  if (typeof expr !== "string" || expr.length === 0) {
    return "schedule must be a non-empty string";
  }
  const trimmed = expr.trim();
  if (trimmed in SHORTCUTS) return null;

  const everyMatch = trimmed.match(EVERY_DURATION);
  if (everyMatch?.groups) {
    const n = Number(everyMatch.groups.n);
    const unit = everyMatch.groups.unit?.toLowerCase();
    if (!Number.isInteger(n) || n <= 0) return "@every duration must be a positive integer";
    const seconds = unit === "s" ? n : unit === "m" ? n * 60 : n * 3600;
    if (seconds < 60) return "@every duration must be at least 60 seconds (v1 resolution floor)";
    return null;
  }

  const fields = trimmed.split(/\s+/);
  if (fields.length !== 5) {
    return `schedule must be 5 cron fields or a documented shortcut, got ${fields.length} field(s)`;
  }
  try {
    CronExpressionParser.parse(trimmed);
    return null;
  } catch (e) {
    return `invalid cron expression: ${(e as Error).message}`;
  }
}

File: go/internal/manifest/cron.go
package manifest

import (
	"fmt"
	"strconv"
	"strings"

	"github.com/robfig/cron/v3"
)

var shortcuts = map[string]struct{}{
	"@hourly":   {},
	"@daily":    {},
	"@midnight": {},
	"@weekly":   {},
	"@monthly":  {},
	"@yearly":   {},
	"@annually": {},
}

var stdCron = cron.NewParser(cron.Minute | cron.Hour | cron.Dom | cron.Month | cron.Dow)

// validateSchedule returns an empty string if the schedule expression is
// well-formed, otherwise a human-readable error message. The accepted forms
// are 5-field cron, the documented shortcuts (D-004), and `@every <N>{s|m|h}`
// where N seconds resolve to ≥ 60.
func validateSchedule(expr string) string {
	if expr == "" {
		return "schedule must be a non-empty string"
	}
	t := strings.TrimSpace(expr)
	if _, ok := shortcuts[t]; ok {
		return ""
	}
	if rest, ok := strings.CutPrefix(t, "@every"); ok {
		rest = strings.TrimSpace(rest)
		if rest == "" || len(rest) < 2 {
			return "@every duration must be a positive integer with unit s|m|h"
		}
		unit := rest[len(rest)-1]
		num, err := strconv.Atoi(rest[:len(rest)-1])
		if err != nil || num <= 0 {
			return "@every duration must be a positive integer"
		}
		var seconds int
		switch unit {
		case 's', 'S':
			seconds = num
		case 'm', 'M':
			seconds = num * 60
		case 'h', 'H':
			seconds = num * 3600
		default:
			return "@every duration unit must be s, m, or h"
		}
		if seconds < 60 {
			return "@every duration must be at least 60 seconds (v1 resolution floor)"
		}
		return ""
	}
	fields := strings.Fields(t)
	if len(fields) != 5 {
		return fmt.Sprintf("schedule must be 5 cron fields or a documented shortcut, got %d field(s)", len(fields))
	}
	if _, err := stdCron.Parse(t); err != nil {
		return fmt.Sprintf("invalid cron expression: %v", err)
	}
	return ""
}
